File: gui/iode_objs/misc/dyn_adjustment.cpp
#include "dyn_adjustment.h"


DynAdjustmentDialog::DynAdjustmentDialog(QString& lec, QWidget* parent) : 
    IodeSettingsDialog(parent), lec(lec)
{
	setupUi(this);

    textEdit_equation->setText(lec);

    QList<QString> list_methods;
    for(const std::string& method : v_adjustment_method) list_methods.append(QString::fromStdString(method));
    comboMethod = new WrapperComboBox(label_method->text(), *comboBox_method, REQUIRED_FIELD, list_methods);
    lineCoef1 = new WrapperQLineEdit(label_coefs->text(), *lineEdit_coef1, REQUIRED_FIELD);
    lineCoef2 = new WrapperQLineEdit(label_coefs->text(), *lineEdit_coef2, REQUIRED_FIELD);

    mapFields["Method"] = comboMethod;
    mapFields["Coef1"] = lineCoef1;
    mapFields["Coef2"] = lineCoef2;

	className = "EQUATION_DYNAMIC_ADJUSTMENT";
    loadSettings();
}

DynAdjustmentDialog::~DynAdjustmentDialog()
{
    delete comboMethod;
    delete lineCoef1;
    delete lineCoef2;
}

void DynAdjustmentDialog::adjust()
{
    try
    {
        std::string std_lec = textEdit_equation->toPlainText().toStdString();
        IodeAdjustmentMethod method = (IodeAdjustmentMethod) comboMethod->extractAndVerify();
        std::string coef1 = lineCoef1->extractAndVerify().toStdString();
        std::string coef2 = lineCoef2->extractAndVerify().toStdString();

        std::string adjusted_lec = dynamic_adjustment(method, std_lec, coef1, coef2);
        lec = QString::fromStdString(adjusted_lec);

        this->accept();
    }
    catch(const std::exception& e)
    {
        QMessageBox::warning(nullptr, "WARNING", QString(e.what()));
    }
}

void DynAdjustmentDialog::help()
{
	QDesktopServices::openUrl(url_manual);
}

File: gui/iode_gui/iode_objs/misc/dyn_adjustment.py
from qtpy.QtWidgets import QMessageBox
from qtpy.QtCore import Slot

from iode_gui.settings import MixinSettingsDialog
from .ui_dyn_adjustment import Ui_DynAdjustmentDialog

from iode import AdjustmentMethod, dynamic_adjustment


class DynAdjustmentDialog(MixinSettingsDialog):

    def __init__(self, lec: str, parent=None):
        super().__init__(parent)
        self.lec: str = lec

        self.ui = Ui_DynAdjustmentDialog()
        self.ui.setupUi(self)

        self.ui.textEdit_equation.setText(lec)
        self.v_adjustment_methods = list(AdjustmentMethod)
        self.ui.comboBox_method.addItems([method.name for method in self.v_adjustment_methods])

    @Slot()
    def adjust(self):
        try:
            lec = self.ui.textEdit_equation.toPlainText()
            method_index = self.ui.comboBox_method.currentIndex()
            method = self.v_adjustment_methods[method_index]
            coef1 = self.ui.lineEdit_coef1.text()
            coef2 = self.ui.lineEdit_coef2.text()

            self.lec = dynamic_adjustment(method, lec, coef1, coef2)
            self.accept()
        except Exception as e:
            QMessageBox.warning(None, "WARNING", str(e))
